src/omni_epd/displays/inky_display.py

```diff
@@ -33,123 +34,77 @@ class InkyDisplay(VirtualEPD):
-    """
-
-    pkg_name = INKY_PKG
-    mode = 'color'  # this uses color by default
-    max_colors = 8  # 7 + CLEAN (no color)
-    modes_available = ('bw', 'color')
-
-    def __init__(self, deviceName, config):
-        super().__init__(deviceName, config)
-
-        # load the device driver
-        deviceObj = self.load_display_driver(self.pkg_name, 'inky_uc8159')
-        self._device = deviceObj.Inky()
-
-        # set the width and height
-        self.width = self._device.width
-        self.height = self._device.height
-
-        # get colors from the inky lib (won't be used normally as inky does conversion)
-        self.palette_filter = deviceObj.DESATURATED_PALETTE
+    def load_device(self, deviceName):
+        # need to figure out what type of device we have
+        dType, dColor, *_ = deviceName.split('_') + [None]
+        if(dType == 'phat'):
+            deviceObj = self.load_display_driver(self.pkg_name, 'phat')
+            device = deviceObj.InkyPHAT(dColor)
+        elif(dType == 'phat1608'):
+            deviceObj = self.load_display_driver(self.pkg_name, 'phat')
+            device = deviceObj.InkyPHAT_SSD1608(dColor)
+        elif(dType == 'what'):
+            deviceObj = self.load_display_driver(self.pkg_name, 'what')
+            device = deviceObj.InkyWHAT(dColor)
+        elif(dType == 'impression'):
+            deviceObj = self.load_display_driver(self.pkg_name, 'inky_uc8159')
+            device = deviceObj.Inky()
+        elif(dType == 'auto'):
+            deviceObj = self.load_display_driver(self.pkg_name, 'auto')
+            device = deviceObj.auto()
+
+        if(device.colour == 'multi'):
+            return device, device.CLEAN, 'color'
+        else:
+            return device, device.WHITE, dColor
 
     @staticmethod
     def get_supported_devices():
-        result = []
-
-        # python libs for this might not be installed - that's ok, return nothing
-        if(check_module_installed(INKY_PKG)):
-            # if passed return list of devices
-            result = [f"{INKY_PKG}.impression"]
-
-        return result
+        return [] if not check_module_installed(INKY_PKG) else [f"{INKY_PKG}.{n}" for n in InkyDisplay.deviceList]
 
+    # set the image and display
     def _display(self, image):
+        # set border
+        self._device.set_border(getattr(self._device, self._get_device_option('border', '').upper(), self._device.border_colour))
 
-        # no palette adjustments when color as the Inky lib does them from the image
-        if(self.mode == 'bw'):
+        # apply any needed conversions to this image based on the mode
+        if(self.mode == 'color'):
+            saturation = self._getfloat_device_option('saturation', .5)  # .5 is default from Inky lib
+            self._device.set_image(image.convert("RGB"), saturation=saturation)
+        else:
             image = self._filterImage(image)
+            self._device.set_image(image.convert("RGB"))

```

Comment:
Is there a specific reason you did the conversion back to RGB here? In the [current version](https://github.com/robweber/omni-epd/blob/main/src/omni_epd/displays/inky_display.py#L90) for non-Impression devices this conversion doesn't happen. On my pHat device using the `convert` function causes the Inky driver to throw an error. If I remove the conversion it works. Happens in both `bw` and `red` mode on that device. I do see it was part of the Impression driver so I imagine it's required for that device. I think checking for `color` mode and doing the conversion based on that should work. 

I can remove this after merge but wanted to check if there was a particular reason why it was there. Below is the error I get. Removing the convert got it going again. 

```
 File "/home/pi/omni-epd-test/src/omni_epd/displays/inky_display.py", line 103, in _display
    self._device.set_image(image.convert("RGB"))
  File "/home/pi/.local/lib/python3.9/site-packages/inky/inky.py", line 368, in set_image
    self.buf = numpy.array(image, dtype=numpy.uint8).reshape((self.height, self.width))
ValueError: cannot reshape array of size 66144 into shape (104,212)

```

Reply:
I was getting inverted colors in bw mode. Maybe only do it on the impression by checking `device.colour == 'multi'` instead.

Reply:
Were you inverted colors before or after the fix for #64? I just pushed that this morning as I was getting inverted BW even in mainline due to that fix. This branch isn't up to date with that fix. 

If you are let me know and I'll revisit that patch.

Reply:
Before. I just checked, and it still works right.

Reply:
like `self._device`? :) i've added this suggestion to the branch, will give it some testing now

Reply:
Oh yeah, they're the same, aren't they? :P

Reply:
It should amount to basically the same check. On[ line 83](https://github.com/robweber/omni-epd/pull/60/files#diff-da6ba08c6a5418bb22e22d387fefc29c37a71c62871e07cf1d05b617e09e26b8R83) it's checking for `multi` when the driver is loaded and then setting the class `mode` variable to `color`. Checking for that is pretty much the same check as `multi` at that point.

Reply:
red inkywhat and inky impresssion both working fine in bw mode

![20220304_210648](https://user-images.githubusercontent.com/413860/156841592-66faeab9-eb01-442f-b66e-d5705ada45db.jpg)

Reply:
also working for red / color mode
![20220304_210911](https://user-images.githubusercontent.com/413860/156841878-f907b0d4-00c0-48bd-81eb-6025ce6ed5cd.jpg)
s

Reply:
It's [setting](https://github.com/robweber/omni-epd/blob/ce01b743f0bc4eba114baa672b4c2b56b7affff2/src/omni_epd/displays/inky_display.py#L45) `dColor` to 'color', not `mode`. Inky impression could also be set to `bw`, whereas it's `device.colour` will always be 'multi'.

It needs to always convert to RGB so that the driver can [convert it](https://github.com/pimoroni/inky/blob/7504410c764f12e1e908cb4a6263aadabe6343a2/library/inky/inky_uc8159.py#L374) to the right palette.

Reply:
Yup, you're right. I see where you're going with that, it's more about checking the device than the mode set by the user. In that case the device `colour` is more accurate.